s2a/src/test/java/io/grpc/s2a/handshaker/IntegrationTest.java

```diff
@@ -0,0 +1,322 @@
+    mtlsS2AServer.shutdown();
+  }
+
+  @Test
+  public void clientCommunicateUsingS2ACredentials_succeeds() throws Exception {
+    ExecutorService executor = Executors.newSingleThreadExecutor();
+    ChannelCredentials credentials =
+        S2AChannelCredentials.createBuilder(s2aAddress).setLocalSpiffeId("test-spiffe-id").build();
+    ManagedChannel channel =
+        Grpc.newChannelBuilder(serverAddress, credentials).executor(executor).build();
+
+    assertThat(doUnaryRpc(executor, channel)).isTrue();
+  }
+
+  @Test
+  public void clientCommunicateUsingS2ACredentialsNoLocalIdentity_succeeds() throws Exception {
+    ExecutorService executor = Executors.newSingleThreadExecutor();
+    ChannelCredentials credentials = S2AChannelCredentials.createBuilder(s2aAddress).build();
+    ManagedChannel channel =
+        Grpc.newChannelBuilder(serverAddress, credentials).executor(executor).build();
+
+    assertThat(doUnaryRpc(executor, channel)).isTrue();
+  }
+
+  @Test
+  public void clientCommunicateUsingMtlsToS2ACredentials_succeeds() throws Exception {
+    ExecutorService executor = Executors.newSingleThreadExecutor();
+    ChannelCredentials credentials =
+        MtlsToS2AChannelCredentials.createBuilder(
+                /* s2aAddress= */ mtlsS2AAddress,
+                /* privateKeyPath= */ "src/test/resources/client_key.pem",
+                /* certChainPath= */ "src/test/resources/client_cert.pem",
+                /* trustBundlePath= */ "src/test/resources/root_cert.pem")
+            .build()
+            .setLocalSpiffeId("test-spiffe-id")
+            .build();
+    ManagedChannel channel =
+        Grpc.newChannelBuilder(serverAddress, credentials).executor(executor).build();
+
+    assertThat(doUnaryRpc(executor, channel)).isTrue();
+  }
+
+  @Test
+  public void clientCommunicateUsingS2ACredentials_s2AdelayStart_succeeds() throws Exception {
+    DoUnaryRpc doUnaryRpc = new DoUnaryRpc();
+    doUnaryRpc.start();
+    Thread.sleep(2000);
+    s2aDelayServer.start();
+    doUnaryRpc.join();
+  }
+
+  private class DoUnaryRpc extends Thread {
+    @Override
+    public void run() {
+      ExecutorService executor = Executors.newSingleThreadExecutor();
+      ChannelCredentials credentials = S2AChannelCredentials.createBuilder(s2aDelayAddress).build();
+      ManagedChannel channel =
+          Grpc.newChannelBuilder(serverAddress, credentials).executor(executor).build();
+      boolean result = false;
+      try {
+        result = doUnaryRpc(executor, channel);
+      } catch (InterruptedException e) {
+        logger.log(Level.SEVERE, "Failed to do unary rpc", e);
+        result = false;
+      }
+      assertThat(result).isTrue();
+    }
+  }
+
+  public static boolean doUnaryRpc(ExecutorService executor, ManagedChannel channel)
+      throws InterruptedException {
+    try {
+      SimpleServiceGrpc.SimpleServiceBlockingStub stub =
+          SimpleServiceGrpc.newBlockingStub(channel).withWaitForReady();
```

Comment:
You should remove the waitForReady.  (As an aside, anytime you use waitForReady you need to add a deadline)

Reply:
Done in https://github.com/grpc/grpc-java/pull/11113/commits/0e059e4d1a3a4477abe53f7d9fd9f411a9395990